diff -uwrN --exclude=.svn nethack-3.4.3/include/config.h nethack-autoconf/include/config.h
--- nethack-3.4.3/include/config.h	2003-12-08 00:39:13.000000000 +0100
+++ nethack-autoconf/include/config.h	2008-11-20 14:27:58.000000000 +0100
@@ -37,12 +37,16 @@
 
 #include "config1.h"	/* should auto-detect MSDOS, MAC, AMIGA, and WIN32 */
 
+#ifdef AUTOCONF
+# include "autoconf.h"
+#endif
+
 
 /* Windowing systems...
  * Define all of those you want supported in your binary.
  * Some combinations make no sense.  See the installation document.
  */
-#define TTY_GRAPHICS	/* good old tty based graphics */
+/* #define TTY_GRAPHICS */	/* good old tty based graphics */
 /* #define X11_GRAPHICS */	/* X11 interface */
 /* #define QT_GRAPHICS */	/* Qt interface */
 /* #define GNOME_GRAPHICS */	/* Gnome interface */
@@ -171,6 +175,7 @@
  *	compression.
  */
 
+#ifndef AUTOCONF
 #ifdef UNIX
 /* path and file name extension for compression program */
 //#define COMPRESS "/usr/bin/compress"	/* Lempel-Ziv compression */
@@ -190,6 +195,7 @@
  *	for detailed configuration.
  */
 #define DLB /* not supported on all platforms */
+#endif /* AUTOCONF */
 
 /*
  *	Defining INSURANCE slows down level changes, but allows games that
@@ -321,6 +327,7 @@
  * complexity of the game but also to the size of the load module.
  */
 
+#ifndef AUTOCONF
 /* dungeon features */
 #define SINKS		/* Kitchen sinks - Janet Walz */
 /* dungeon levels */
@@ -339,14 +346,17 @@
 # define CLIPPING	/* allow smaller screens -- ERS */
 #endif
 
+#define EXP_ON_BOTL	/* Show experience on bottom line */
+/* #define SCORE_ON_BOTL */	/* added by Gary Erickson (erickson@ucivax) */
+
+#define REALTIME_ON_BOTL	/* Show elapsed time on bottom line.  Note:
+                                 * this breaks savefile compatibility. */
+#endif /* AUTOCONF */
+
 #ifdef REDO
 # define DOAGAIN '\001' /* ^A, the "redo" key used in cmd.c and getline.c */
 #endif
 
-#define EXP_ON_BOTL	/* Show experience on bottom line */
-#define SCORE_ON_BOTL /* added by Gary Erickson (erickson@ucivax) */
-#define REALTIME_ON_BOTL	/* Show elapsed time on bottom line.  Note:
-                                 * this breaks savefile compatibility. */
 /* The options in this section require the extended logfile support */
 #ifdef XLOGFILE
 #define RECORD_CONDUCT  /* Record conducts kept in logfile */
diff -uwrN --exclude=.svn nethack-3.4.3/include/ntconf.h nethack-autoconf/include/ntconf.h
--- nethack-3.4.3/include/ntconf.h	2003-12-08 00:39:13.000000000 +0100
+++ nethack-autoconf/include/ntconf.h	2008-09-25 20:15:24.000000000 +0200
@@ -5,7 +5,7 @@
 #ifndef NTCONF_H
 #define NTCONF_H
 
-/* #define SHELL	/* nt use of pcsys routines caused a hang */
+/* #define SHELL */	/* nt use of pcsys routines caused a hang */
 
 #define RANDOM		/* have Berkeley random(3) */
 #define TEXTCOLOR	/* Color text */
@@ -23,13 +23,13 @@
 
 #define SELF_RECOVER		/* Allow the game itself to recover from an aborted game */
 
-#define USER_SOUNDS
+//#define USER_SOUNDS
 /*
  * -----------------------------------------------------------------
  *  The remaining code shouldn't need modification.
  * -----------------------------------------------------------------
  */
-/* #define SHORT_FILENAMES	/* All NT filesystems support long names now */
+/* #define SHORT_FILENAMES */	/* All NT filesystems support long names now */
 
 #ifdef MICRO
 #undef MICRO			/* never define this! */
@@ -49,7 +49,7 @@
 #ifdef MSWIN_GRAPHICS
 #define OPTIONS_USED	"guioptions"
 #else
-#define OPTIONS_USED	"ttyoptions"
+#define OPTIONS_USED	"options"
 #endif
 #define OPTIONS_FILE OPTIONS_USED
 
diff -uwrN --exclude=.svn nethack-3.4.3/include/pcconf.h nethack-autoconf/include/pcconf.h
--- nethack-3.4.3/include/pcconf.h	2003-12-08 00:39:13.000000000 +0100
+++ nethack-autoconf/include/pcconf.h	2008-09-23 22:44:01.000000000 +0200
@@ -133,7 +133,9 @@
  * simply a flag to turn on or off napping for visual effects at run-time.
  */
 
+#ifndef AUTOCONF
 #define TIMED_DELAY	/* enable the `timed_delay' run-time option */
+#endif
 
 # ifdef PCMUSIC
 #define TIMED_DELAY	/* need it anyway */
diff -uwrN --exclude=.svn nethack-3.4.3/include/system.h nethack-autoconf/include/system.h
--- nethack-3.4.3/include/system.h	2003-12-08 00:39:13.000000000 +0100
+++ nethack-autoconf/include/system.h	2008-09-21 15:54:26.000000000 +0200
@@ -9,6 +9,21 @@
 
 #define E extern
 
+#ifdef AUTOCONF
+/*
+ * When using GNU autoconf to configure, the switches are output to
+ * autoconf.h which should override the defaults set earlier in
+ * config.h and *conf.h
+ */
+#include "autoconf.h"
+#if HAVE_STRING_H
+#include <string.h>
+#endif
+#if HAVE_DECL_STRNICMP && !HAVE_DECL_STRNCMPI
+#define strncmpi(a,b,c) strnicmp(a,b,c)
+#endif
+#endif
+
 /* some old <sys/types.h> may not define off_t and size_t; if your system is
  * one of these, define them by hand below
  */
diff -uwrN --exclude=.svn nethack-3.4.3/include/unixconf.h nethack-autoconf/include/unixconf.h
--- nethack-3.4.3/include/unixconf.h	2003-12-08 00:39:13.000000000 +0100
+++ nethack-autoconf/include/unixconf.h	2008-09-23 22:24:57.000000000 +0200
@@ -132,8 +132,10 @@
  * "extra output" method is used, but not all systems provide access to
  * a fine-grained timer.
  */
+#ifndef AUTOCONF
 /* #define TIMED_DELAY */	/* usleep() */
 #endif
+#endif
 
 /*
  * If you define MAIL, then the player will be notified of new mail
@@ -264,6 +266,8 @@
 
 #define HLOCK	"perm"	/* an empty file used for locking purposes */
 
+#include "system.h"
+
 #ifndef REDO
 #define Getchar nhgetch
 #endif
@@ -271,8 +275,6 @@
 
 #define SHELL		/* do not delete the '!' command */
 
-#include "system.h"
-
 #if defined(POSIX_TYPES) || defined(__GNUC__)
 #include <stdlib.h>
 #include <unistd.h>
diff -uwrN --exclude=.svn nethack-3.4.3/win/tty/termcap.c nethack-autoconf/win/tty/termcap.c
--- nethack-3.4.3/win/tty/termcap.c	2003-12-08 00:39:14.000000000 +0100
+++ nethack-autoconf/win/tty/termcap.c	2008-09-23 21:28:49.000000000 +0200
@@ -835,7 +835,7 @@
 
 #include <curses.h>
 
-#ifndef LINUX
+#ifndef HAVE_TPARM
 extern char *tparm();
 #endif
 
diff -uwrN --exclude=.svn nethack-3.4.3/win/win32/winhack.c nethack-autoconf/win/win32/winhack.c
--- nethack-3.4.3/win/win32/winhack.c	2003-12-08 00:39:14.000000000 +0100
+++ nethack-autoconf/win/win32/winhack.c	2008-09-23 21:28:49.000000000 +0200
@@ -12,7 +12,7 @@
 #include "mhmain.h"
 #include "mhmap.h"
 
-#ifndef __BORLANDC__
+#if !defined(__BORLANDC__) && !defined(__MINGW32__)
 #include <shlwapi.h>
 #else /* Borland redefines "boolean" in shlwapi.h so just use the little bit we need */
 typedef struct _DLLVERSIONINFO
