--- config.h	Sat Aug 30 02:07:15 2003
+++ ../../nethack-3.4.0/include/config.h	Wed Nov 19 20:34:40 2003
@@ -1,4 +1,4 @@
-/*	SCCS Id: @(#)config.h	3.4	2003/02/19	*/
+/*	SCCS Id: @(#)config.h	3.4	2002/03/17	*/
 /* Copyright (c) Stichting Mathematisch Centrum, Amsterdam, 1985. */
 /* NetHack may be freely redistributed.  See license for details. */
 
@@ -83,9 +83,7 @@
 #endif
 
 #ifdef QT_GRAPHICS
-# define DEFAULT_WC_TILED_MAP   /* Default to tiles if users doesn't say wc_ascii_map */
 # define USER_SOUNDS		/* Use sounds */
-# define USER_SOUNDS_REGEX
 # define USE_XPM		/* Use XPM format for images (required) */
 # define GRAPHIC_TOMBSTONE	/* Use graphical tombstone (rip.ppm) */
 # ifndef DEFAULT_WINDOW_SYS
@@ -133,10 +131,9 @@
 
 /*
  * Section 2:	Some global parameters and filenames.
- *		Commenting out WIZARD, LOGFILE, NEWS or PANICLOG removes that
- *		feature from the game; otherwise set the appropriate wizard
- *		name.  LOGFILE, NEWS and PANICLOG refer to files in the
- *		playground.
+ *		Commenting out WIZARD, LOGFILE, or NEWS removes that feature
+ *		from the game; otherwise set the appropriate wizard name.
+ *		LOGFILE and NEWS refer to files in the playground.
  */
 
 #ifndef WIZARD		/* allow for compile-time or Makefile changes */
@@ -150,7 +147,6 @@
 
 #define LOGFILE "logfile"	/* larger file for debugging purposes */
 #define NEWS "news"		/* the file containing the latest hack news */
-#define PANICLOG "paniclog"	/* log of panic and impossible events */
 
 /*
  *	If COMPRESS is defined, it should contain the full path name of your
@@ -167,8 +163,8 @@
 
 #ifdef UNIX
 /* path and file name extension for compression program */
-#define COMPRESS "/usr/bin/compress"	/* Lempel-Ziv compression */
-#define COMPRESS_EXTENSION ".Z"		/* compress's extension */
+#define COMPRESS "/usr/bin/bzip2"	/* Lempel-Ziv compression */
+#define COMPRESS_EXTENSION ".bz"		/* compress's extension */
 /* An example of one alternative you might want to use: */
 /* #define COMPRESS "/usr/local/bin/gzip" */	/* FSF gzip compression */
 /* #define COMPRESS_EXTENSION ".gz" */		/* normal gzip extension */
@@ -350,7 +346,28 @@
  * bugs left here.
  */
 
+#ifdef TTY_GRAPHICS
+# define MENU_COLOR
+#endif
+#define USE_REGEX_MATCH
+/* if USE_REGEX_MATCH is defined, use regular expressions (GNU regex.h)
+ * otherwise use pmatch() to match menu color lines.
+ * pmatch() provides basic globbing: '*' and '?' wildcards.
+ */
+#define FLIPCOIN
 /*#define GOLDOBJ */	/* Gold is kept on obj chains - Helge Hafting */
+#define DUMP_LOG
+#define DEATH_EXPLORE 
+#define PARANOID 
+#define SHOW_BORN 
+#define LOG_MOVES
+
+
+
+/*
+ * OUTPUT SCREEN & STATES DUMP of HTML
+ */
+#define HTMLDUMP
 
 /* End of Section 5 */
 
